lib/AST/RequirementMachine/Symbol.cpp

```diff
@@ -80,6 +80,14 @@ struct Symbol::Storage final
     Kind = Kind::Shape;
   }
 
+  /// A dummy type for overload resolution of the
+  /// 'pack element' constructor for Storage.
+  struct ForPackElement {};
```

Comment:
I think you can get rid of both ForPack and ForShape and just add a Storage(Kind) overload, but it's a matter of taste.